File: newfolder/II/2024_2025_II_Kharkiv/2024_Kharkiv_7.py
# G (358) Квадрати
# Всеукраїнська олімпіада з інформатики ІІ етап (Kharkiv) 2024-2025 н.р.
# https://uoi.eolymp.space/uk/problems/358
# https://www.kievoi.ippo.kubg.edu.ua/kievoi/1-2/2024.pdf умова
# https://www.youtube.com/watch?v=QOVWCRvo4dE розбір (лише A-F)


# 65% (перевищено ліміт часу - 0,2 с) на eolymp
# Коротке замикання дасть суттєву економію в часі для випадків,
# де кількість 0 і 1 є приблизно однаковим

n = int(input())
N = [list(map(int, input().split())) for _ in range(n)]
ans = 0
# опорна (перша) точка квадрата має координати (i,j). Наступна точка віддалена
# на r рядків вниз і c стовпчиків вправо
# від першої точки  (i,j)
for r in range(n-1):
    for c in range(1, n):
        # досліджуємо кожну точку
        for i in range(n - c - r):
            for j in range(r, n - c):
                s = N[i][j] + N[i + r][j + c] + N[i + c + r][j + c - r] + \
                    N[i + c][j - r]
                if s == 0 or s == 4:
                    # print(i,j, 'r=',r, 'c=',c)
                    ans += 1

print(ans)

# Variant2 з коротким замиканням  і зменшенням кількості операцій +,-
# пройшов тест на рідному порталі цієї задачі на 100%
# https://olympkh.dots.org.ua/
# на eolymp цей варіант дає лише 70% (через перевищення ліміту часу у 0.2 с)

# n = int(input())
# N = [input().split() for _ in range(n)]
# for i in range(n):
#     for j in range(n):
#         N[i][j] = True if N[i][j] == '1' else False
#
# ans = 0
# for c in range(1, n):
#     nc = n - c
#     for i in range(nc):
#         ic = i + c
#         for j in range(nc):
#             jc = j + c
#             if N[i][j] == N[i][jc] and N[i][j] == N[ic][jc] and N[i][j] == \
#                     N[ic][j]:
#                 ans += 1
#
# for r in range(1, n - 1):
#     for c in range(1, n):
#         nj = n - c
#         cmr = c - r
#         cr = c + r
#         for i in range(nj - r):
#             icr = i + cr
#             ir = i + r
#             for j in range(r, nj):
#
#                 if N[i][j] and N[ir][j + c] and N[icr][j + cmr] and N[i + c][j - r]:
#                     ans += 1
#                 if not (N[i][j] or N[ir][j + c] or N[icr][j + cmr] or N[i + c][j - r]):
#                     ans += 1
#
# print(ans)

File: newfolder/II/2024_2025_II_Kharkiv/2024_Kharkiv_7.cpp
// згенeровано в ChatGPT на основі коду на Python
// пройшло всі 100% тестів на eolymp з максимальним часом 6 мс,
// тобто в 50 разів швидше, ніж код з такою ж структурою на Python
#include <iostream>
#include <vector>
#include <string>

int main() {
    int n;
    std::cin >> n;

    // Зчитування вхідних даних
    std::vector<std::vector<std::string>> N(n, std::vector<std::string>(n));
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            std::cin >> N[i][j];
        }
    }

    int ans = 0;

    // Перша частина: перевірка квадратів
    for (int c = 1; c < n; ++c) {
        int nc = n - c;
        for (int i = 0; i < nc; ++i) {
            int ic = i + c;
            for (int j = 0; j < nc; ++j) {
                int jc = j + c;
                if (N[i][j] == N[i][jc] && N[i][j] == N[ic][jc] && N[i][j] == N[ic][j]) {
                    ans++;
                }
            }
        }
    }

    // Друга частина: перевірка "ромбів"
    for (int r = 1; r < n - 1; ++r) {
        for (int c = 1; c < n; ++c) {
            int nj = n - c;
            int cmr = c - r;
            int cr = c + r;
            for (int i = 0; i < nj - r; ++i) {
                int icr = i + cr;
                int ir = i + r;
                for (int j = r; j < nj; ++j) {
                    if (N[i][j] == N[ir][j + c] && N[i][j] == N[icr][j + cmr] && N[i][j] == N[i + c][j - r]) {
                        ans++;
                    }
                }
            }
        }
    }

    std::cout << ans << std::endl;
    return 0;
}
